### ISSUE TITLE
Darkside mode crashes when getblockchaininfo is called and there are no blocks loaded.

### ISSUE THREAD

**Issue by pacu** [zcash/lightwalletd#451]
**What is the bug?**
When started, darksidewalletd seems to attempt to connect to zcashd and crashes.

**Additional context**

commit: `8269810eeefac56ce326b0f6878c1ea3821b8ad5` 

I was testing new darkside tests cases for all wallets. It suddenly stopped working. 

here's the console log

```
lightwalletd % sh ~/darksidewalletd.sh  
{"app":"lightwalletd","buildDate":"2023-08-27","buildUser":"pacu","gitCommit":"8269810eeefac56ce326b0f6878c1ea3821b8ad5","level":"info","msg":"Starting gRPC server version v0.4.15-5-g8269810 on 127.0.0.1:9067","time":"2023-08-27T18:33:11-03:00"}
{"app":"lightwalletd","level":"warning","msg":"Starting insecure no-TLS (plaintext) server","time":"2023-08-27T18:33:11-03:00"}
Starting insecure server
{"app":"lightwalletd","level":"info","msg":"Reading 0 blocks (since Sapling activation) from disk cache ...","time":"2023-08-27T18:33:11-03:00"}
{"app":"lightwalletd","level":"info","msg":"Done reading 0 blocks from disk cache","time":"2023-08-27T18:33:11-03:00"}
{"app":"lightwalletd","level":"info","msg":"Darkside mode running","time":"2023-08-27T18:33:11-03:00"}
{"app":"lightwalletd","level":"fatal","msg":"getblockchaininfo: no blocks","time":"2023-08-27T18:33:15-03:00"}
Lightwalletd died with a Fatal error. Check logfile for details.
```

script to start darkside mode

```
#!/bin/zsh

./lightwalletd --log-file /dev/stdout --darkside-very-insecure  --darkside-timeout 1000 --gen-cert-very-insecure --data-dir . --no-tls-very-insecure
```

after it crashed this folder can be found where the binary is at.

<img width="620" alt="image" src="https://github.com/zcash/lightwalletd/assets/484008/3f2d60f6-5041-4351-962a-644f86d44b36">

****
**Solution**
Add fix suggestions here, or what you’d like to see.

**Comment by LarryRuane** [comment]
I'm unable to reproduce this. After starting `lightwalletd` with the arguments listed above, it prints the messages up to and including "Darkside mode running" but then it continues running (waiting for a request), without crashing.

I was able to recreate the exact crash by running:
```
$ grpcurl -plaintext localhost:9067 cash.z.wallet.sdk.rpc.CompactTxStreamer/GetLightdInfo
```
This isn't a regression. It crashes because there is no blockchain information to return.

I could make a code change to not crash, but to return something reasonable instead, if that would be helpful. (Looking at it now, it does seem harsh to crash in this case; I don't recall if I had a strong reason for implementing this behavior, but I don't think so.)

**Comment by pacu** [comment]
Uhmmm. This is really useful though!

I don't recall having any wallets running in the background that could be requesting that RPC 🤨 maybe I had a zombie process of Zingo-cli? 

If you can make it not to crash and return some meaningful error on how not to get into that situation is always for the best.

If we managed to make this an automated test, the crash wouldn't be of much help for someone reading the log. 

Thanks Larry !

**Comment by LarryRuane** [comment]
What's useful (I don't quite understand)?

We've had this question before, and I'm not sure of the best answer. We can make darksidewallet kind of "fragile" (like it is now with respect to this issue), or more forgiving. At first thought, forgiving seems better. But being forgiving may hide some defects in the test framework or even the real production code. Seems like "fail fast" (as opposed to trying to "recover") may be the best option for test code in terms of uncovering bugs, but it's obviously disruptive when things like this happen. It's a tradeoff, and I don't know the best answer. (You can tell me what you prefer.)

To be more specific, in this case, it's not like `lightwalletd` just dies with no clue about what happened; there is an error message in the log file:
```
{"app":"lightwalletd","level":"fatal","msg":"getblockchaininfo: no blocks","time":"2023-08-27T16:38:58-06:00"}
```
I think this error is fairly clear (the darkside server is receiving a `getblockchaininfo` request when there are no blocks).

But what do you think? Should this gRPC succeed if no blocks exist?

If so, then let's look at the `GetLightdInfo` output after creating some blocks (on darksidewallet):
```
$ grpcurl -plaintext localhost:9067 cash.z.wallet.sdk.rpc.CompactTxStreamer/GetLightdInfo
{
  "version": "v0.4.16-3-g8269810",
  "vendor": "ECC DarksideWalletD",
  "taddrSupport": true,
  "chainName": "x",
  "saplingActivationHeight": "663190",
  "consensusBranchId": "bad",
  "blockHeight": "663190",
  "gitCommit": "8269810eeefac56ce326b0f6878c1ea3821b8ad5",
  "buildDate": "2023-08-27",
  "buildUser": "larry"
}
```
We get the `saplingActivationHeight` and `blockHeight` from the blocks that have been created. What values should these fields have if no blocks exist yet? Would zero be okay? Or maybe -1 (definitely an impossible value)?

**Comment by LarryRuane** [comment]
Just want to make it clear that the title of this issue is incorrect -- `lightwalletd` (in darksidewallet mode) is not trying to connect to `zcashd`.

**Comment by pacu** [comment]
> But what do you think? Should this gRPC succeed if no blocks exist?

I think it should definitely should fail. I wonder if we could rephrase the error to make it obvious that it's a user (dev) error and what dlwd would expect instead. I'll think of something and propose it on a PR. what do you thing?

**Comment by pacu** [comment]
I'm testing Zingo-cli and this is the error that's thrown when a TreeState is required but not available in dwld.

This is how it gets shown on the CLI

```
 regtest network in use but no regtest flag recognized!
(regtest) Block:206 (type 'help') >> thread 'tokio-runtime-worker' panicked 
at 'called `Result::unwrap()` on an `Err` value: "Error with get_tree_state response 
at http://localhost:9067/: Status { code: Unknown, message: \"there is no TreeState 
Staged for height \\\"305\\\". Stage it using AddTreeState() first\", metadata: MetadataMap 
{ headers: {\"content-type\": \"application/grpc\"} }, source: None }"',
 zingolib/src/blaze/trial_decryptions.rs:135:28
```

I think I could try to do the same for this case, would that be ok?

**Comment by LarryRuane** [comment]
So instead of stopping lightwalletd (exiting the process), I should change lightwalletd return an error similar to this? I could definitely do that.

**Comment by pacu** [comment]
I'm in doubt with that, because, what's the TreeState of synthetic block? 🤯

**Comment by LarryRuane** [comment]
I don't quite understand. Darksidewallet has a method to add a tree state (`AddTreeState`) and the tree state entry that it adds contains both a height and a block hash (as well as the Sapling and Orchard tree states, of course, and a couple of other things). But currently, darkside only supports `GetTreeState` at a given _block height_ argument, but not at a given _block hash_ argument -- which the production `GetTreeState` does support. It would be a simple change to the darkside's version of `GetTreeState` to allow looking up a tree state by block hash (since we have each tree state's block hash) and return it. The argument to `GetTreeState` already support either height or hash, so there's no change to the interface.

It would seem better for darkside mode to support the same interface as production mode, as far as possible. So that's why I'm thinking that this is an improvement worth implementing (especially since it's simple).

What do you think?

Having written all this, I think I'll just go ahead and implement it, even if you don't need it immediately. It will be a surprise and disruptive if it's needed later and isn't there.

**Comment by pacu** [comment]
I agree! I think that it's for the best that darkside counterparts of RPCs that are used in production support the same inputs in both modes.

**Comment by LarryRuane** [comment]
> I agree! I think that it's for the best that darkside counterparts of RPCs that are used in production support the same inputs in both modes.

Fixed now by #453 -- NOTE, not this issue, this was the result of a side comment / observation / request (there is no ticket for 453).
